File: server/projectQueryService.ts
/**
 * PRISM Canonical Project Query Service
 *
 * Single Source of Truth for querying, filtering, ranking, and aggregating
 * projects across Dashboard, Project Table, Analytics, and AI Copilot.
 *
 * Enforces:
 *  - Deterministic risk bands (CRITICAL: 80-100, HIGH: 60-79, MODERATE: 40-59, LOW: 0-39)
 *  - Canonical state classification (Dedicated in-state vs Multi-state corridors)
 *  - Consistent priority ranking from PRISMPriorityEngine
 *  - Full-dataset KPI calculations (never on paginated slices)
 *  - Zero hallucination / zero frontend risk recalculation
 */

import { Project, RiskTier, PriorityTier, PortfolioKPIs } from '../src/types/index';
import { paimanaRepository } from './paimanaRepository';

export interface CanonicalFilterOptions {
  search?: string;
  state?: string;
  includeMultiState?: boolean;
  sector?: string;
  agency?: string;
  riskTier?: RiskTier | 'ALL';
  priorityTier?: PriorityTier | 'ALL';
  sortBy?: string;
  sortDirection?: 'asc' | 'desc';
  limit?: number;
  offset?: number;
}

export interface StatePartition {
  canonicalState: string;
  dedicatedProjects: Project[];
  multiStateProjects: Project[];
  allAssociatedProjects: Project[];
  dedicatedCount: number;
  multiStateCount: number;
  totalAssociatedCount: number;
  dedicatedKPIs: PortfolioKPIs;
  allAssociatedKPIs: PortfolioKPIs;
  dedicatedByRisk: Record<RiskTier | 'UNRATED', number>;
  multiStateByRisk: Record<RiskTier | 'UNRATED', number>;
  totalByRisk: Record<RiskTier | 'UNRATED', number>;
}

export interface CanonicalQueryResult {
  projects: Project[];
  allMatching: Project[];
  totalCount: number;
  kpis: PortfolioKPIs;
}

export class ProjectQueryService {
  /**
   * Universal project filter and query executor.
   * Shared by /api/projects, Dashboard, and Copilot Intent Gate.
   */
  static query(options: CanonicalFilterOptions = {}): CanonicalQueryResult {
    const defaultSortBy = options.sortBy || (options.riskTier && options.riskTier !== 'ALL' ? 'riskScore' : options.priorityTier && options.priorityTier !== 'ALL' ? 'priorityScore' : 'id');
    const defaultSortDir = options.sortDirection || ((defaultSortBy === 'riskScore' || defaultSortBy === 'priorityScore') ? 'desc' : 'asc');

    const { projects: paginated, allMatching, totalCount } = paimanaRepository.listProjects({
      search: options.search,
      state: options.state,
      agency: options.agency,
      sector: options.sector,
      riskTier: options.riskTier,
      priorityTier: options.priorityTier,
      sortBy: defaultSortBy,
      sortDirection: defaultSortDir,
      limit: options.limit,
      offset: options.offset
    });

    // Compute KPIs on the complete matching dataset, NEVER on paginated slice
    const kpis = paimanaRepository.computeKPIs(allMatching);

    return {
      projects: paginated,
      allMatching,
      totalCount,
      kpis
    };
  }

  /**
   * Canonical State Partition & Breakdown.
   * Distinguishes Dedicated in-state projects from Multi-state corridors involving the state.
   */
  static getStatePartition(stateName: string): StatePartition | null {
    const s = stateName.trim().toLowerCase();
    const all = paimanaRepository.listProjects().allMatching;

    // Find canonical name if available
    const exactMatch = all.find(p => p.state.trim().toLowerCase() === s);
    const canonicalState = exactMatch ? exactMatch.state.trim() : stateName.trim();

    const dedicatedProjects = all.filter(p => p.state.trim().toLowerCase() === s);
    const multiStateProjects = all.filter(p =>
      p.state.trim().toLowerCase() !== s &&
      p.state.toLowerCase().includes(s)
    );

    if (dedicatedProjects.length === 0 && multiStateProjects.length === 0) {
      return null;
    }

    const allAssociatedProjects = [...dedicatedProjects, ...multiStateProjects];

    const initRiskCounts = (): Record<RiskTier | 'UNRATED', number> => ({
      CRITICAL: 0,
      HIGH: 0,
      MODERATE: 0,
      LOW: 0,
      UNRATED: 0
    });

    const dedicatedByRisk = initRiskCounts();
    for (const p of dedicatedProjects) {
      const t = (p.riskTier as RiskTier) || 'UNRATED';
      dedicatedByRisk[t] = (dedicatedByRisk[t] || 0) + 1;
    }

    const multiStateByRisk = initRiskCounts();
    for (const p of multiStateProjects) {
      const t = (p.riskTier as RiskTier) || 'UNRATED';
      multiStateByRisk[t] = (multiStateByRisk[t] || 0) + 1;
    }

    const totalByRisk = initRiskCounts();
    for (const p of allAssociatedProjects) {
      const t = (p.riskTier as RiskTier) || 'UNRATED';
      totalByRisk[t] = (totalByRisk[t] || 0) + 1;
    }

    return {
      canonicalState,
      dedicatedProjects,
      multiStateProjects,
      allAssociatedProjects,
      dedicatedCount: dedicatedProjects.length,
      multiStateCount: multiStateProjects.length,
      totalAssociatedCount: allAssociatedProjects.length,
      dedicatedKPIs: paimanaRepository.computeKPIs(dedicatedProjects),
      allAssociatedKPIs: paimanaRepository.computeKPIs(allAssociatedProjects),
      dedicatedByRisk,
      multiStateByRisk,
      totalByRisk
    };
  }

  /**
   * Canonical State + Risk Tier Query.
   * Returns exact matching projects, guaranteed identical to Dashboard filter.
   */
  static getProjectsByStateAndRisk(
    stateName: string,
    riskTier: RiskTier,
    options: { limit?: number; includeMultiState?: boolean } = {}
  ): {
    canonicalState: string;
    riskTier: RiskTier;
    dedicatedProjects: Project[];
    multiStateProjects: Project[];
    totalCount: number;
    matchingProjects: Project[];
    displayProjects: Project[];
    multiStateNotice?: string;
  } {
    const partition = this.getStatePartition(stateName);
    const targetTier = riskTier.toUpperCase() as RiskTier;

    if (!partition) {
      return {
        canonicalState: stateName,
        riskTier: targetTier,
        dedicatedProjects: [],
        multiStateProjects: [],
        totalCount: 0,
        matchingProjects: [],
        displayProjects: []
      };
    }

    // Sort deterministically: riskScore DESC, id ASC
    const sortFn = (a: Project, b: Project) => {
      const sDiff = (b.riskScore ?? 0) - (a.riskScore ?? 0);
      if (sDiff !== 0) return sDiff;
      return a.id.localeCompare(b.id);
    };

    const dedicatedMatching = partition.dedicatedProjects
      .filter(p => p.riskTier === targetTier)
      .sort(sortFn);

    const multiStateMatching = partition.multiStateProjects
      .filter(p => p.riskTier === targetTier)
      .sort(sortFn);

    // Canonical matching matches the Dashboard: dedicated projects in that state
    const canonicalMatching = options.includeMultiState
      ? [...dedicatedMatching, ...multiStateMatching].sort(sortFn)
      : dedicatedMatching;

    const limit = options.limit || 5;
    const displayProjects = canonicalMatching.slice(0, limit);

    let multiStateNotice: string | undefined;
    if (!options.includeMultiState && multiStateMatching.length > 0) {
      multiStateNotice = `There are also ${multiStateMatching.length} multi-state project(s) involving ${partition.canonicalState} with ${targetTier} risk (${multiStateMatching.map(p => `${p.name.split('(')[0].trim()} [${p.id}]`).join(', ')}).`;
    }

    return {
      canonicalState: partition.canonicalState,
      riskTier: targetTier,
      dedicatedProjects: dedicatedMatching,
      multiStateProjects: multiStateMatching,
      totalCount: dedicatedMatching.length,
      matchingProjects: canonicalMatching,
      displayProjects,
      multiStateNotice
    };
  }

  /**
   * Canonical Priority Queue Query.
   * Rank-ordered strictly by priorityScore DESC, id ASC.
   */
  static getPriorityProjects(options: {
    state?: string;
    sector?: string;
    priorityTier?: PriorityTier;
    limit?: number;
  } = {}): {
    projects: Project[];
    totalMatching: number;
    p1Count: number;
    p2Count: number;
    p3Count: number;
  } {
    let list = paimanaRepository.listProjects().allMatching;

    if (options.state && options.state !== 'ALL') {
      const s = options.state.trim().toLowerCase();
      list = list.filter(p => p.state.toLowerCase() === s);
    }

    if (options.sector && options.sector !== 'ALL') {
      const sec = options.sector.trim().toLowerCase();
      list = list.filter(p => (p.sector as string).toLowerCase() === sec || (p.derivedSector || '').toLowerCase() === sec);
    }

    if (options.priorityTier && (options.priorityTier as string) !== 'ALL') {
      const pt = (options.priorityTier as string).trim().toUpperCase() as PriorityTier;
      list = list.filter(p => p.priorityTier === pt);
    }

    // Deterministic sort: priorityScore DESC, riskScore DESC, id ASC
    list.sort((a, b) => {
      const pDiff = (b.priorityScore ?? 0) - (a.priorityScore ?? 0);
      if (pDiff !== 0) return pDiff;
      const rDiff = (b.riskScore ?? 0) - (a.riskScore ?? 0);
      if (rDiff !== 0) return rDiff;
      return a.id.localeCompare(b.id);
    });

    const p1Count = list.filter(p => p.priorityTier === 'P1').length;
    const p2Count = list.filter(p => p.priorityTier === 'P2').length;
    const p3Count = list.filter(p => p.priorityTier === 'P3').length;

    const limit = options.limit || 5;
    return {
      projects: list.slice(0, limit),
      totalMatching: list.length,
      p1Count,
      p2Count,
      p3Count
    };
  }

  /**
   * Canonical Highest Risk Projects.
   * Rank-ordered strictly by riskScore DESC, id ASC.
   */
  static getTopRiskProjects(options: {
    state?: string;
    sector?: string;
    limit?: number;
  } = {}): {
    projects: Project[];
    totalCritical: number;
    totalHigh: number;
  } {
    let list = paimanaRepository.listProjects().allMatching.filter(p => p.riskScore != null);

    if (options.state && options.state !== 'ALL') {
      const s = options.state.trim().toLowerCase();
      list = list.filter(p => p.state.toLowerCase() === s);
    }

    if (options.sector && options.sector !== 'ALL') {
      const sec = options.sector.trim().toLowerCase();
      list = list.filter(p => (p.sector as string).toLowerCase() === sec || (p.derivedSector || '').toLowerCase() === sec);
    }

    const totalCritical = list.filter(p => p.riskTier === 'CRITICAL').length;
    const totalHigh = list.filter(p => p.riskTier === 'HIGH').length;

    // Deterministic sort: riskScore DESC, id ASC
    list.sort((a, b) => {
      const diff = (b.riskScore ?? 0) - (a.riskScore ?? 0);
      if (diff !== 0) return diff;
      return a.id.localeCompare(b.id);
    });

    const limit = options.limit || 5;
    return {
      projects: list.slice(0, limit),
      totalCritical,
      totalHigh
    };
  }

  /**
   * Canonical Stagnant Projects.
   * Derived strictly from PRISM Progress Stagnation indicators (delta <= 0.5 pp).
   */
  static getStagnantProjects(options: {
    state?: string;
    sector?: string;
    limit?: number;
  } = {}): {
    projects: Project[];
    totalStagnant: number;
  } {
    let list = paimanaRepository.listProjects().allMatching;

    if (options.state && options.state !== 'ALL') {
      const s = options.state.trim().toLowerCase();
      list = list.filter(p => p.state.toLowerCase() === s);
    }

    if (options.sector && options.sector !== 'ALL') {
      const sec = options.sector.trim().toLowerCase();
      list = list.filter(p => (p.sector as string).toLowerCase() === sec || (p.derivedSector || '').toLowerCase() === sec);
    }

    // Filter projects where stagnation is detected
    const stagnant = list.filter(p => {
      const hasStagnantDriver = (p.primaryRiskDriver && p.primaryRiskDriver.toLowerCase().includes('stagnat')) ||
        (p.topRiskDrivers && p.topRiskDrivers.some(d => d.label.toLowerCase().includes('stagnat')));
      const hasStagnantAlert = p.alert && p.alert.alertType === 'Progress Stagnation';
      return Boolean(hasStagnantDriver || hasStagnantAlert);
    });

    // Deterministic sort: riskScore DESC, id ASC
    stagnant.sort((a, b) => {
      const diff = (b.riskScore ?? 0) - (a.riskScore ?? 0);
      if (diff !== 0) return diff;
      return a.id.localeCompare(b.id);
    });

    const limit = options.limit || 5;
    return {
      projects: stagnant.slice(0, limit),
      totalStagnant: stagnant.length
    };
  }

  /**
   * Canonical Healthy / Good Pace Projects (Phase 11).
   * Strictly verifies LOW risk tier (0-39), zero schedule slippage, zero cost overrun,
   * and positive progress velocity. Does NOT equate high progress with good pace.
   */
  static getHealthyPaceProjects(options: {
    state?: string;
    sector?: string;
    limit?: number;
  } = {}): {
    projects: Project[];
    totalHealthy: number;
    qualificationNote: string;
  } {
    let list = paimanaRepository.listProjects().allMatching;

    if (options.state && options.state !== 'ALL') {
      const s = options.state.trim().toLowerCase();
      list = list.filter(p => p.state.toLowerCase() === s);
    }

    if (options.sector && options.sector !== 'ALL') {
      const sec = options.sector.trim().toLowerCase();
      list = list.filter(p => (p.sector as string).toLowerCase() === sec || (p.derivedSector || '').toLowerCase() === sec);
    }

    // Verified healthy progression criteria:
    // 1. LOW risk tier (0-39)
    // 2. Zero schedule slippage (timeOverrunMonths <= 0)
    // 3. Zero cost overrun (costOverrunPercent <= 0)
    // 4. Physical progress > 0
    const pristine = list.filter(p =>
      p.riskTier === 'LOW' &&
      (p.timeOverrunMonths || 0) <= 0 &&
      (p.costOverrunPercent || 0) <= 0 &&
      (p.physicalProgressPercent || 0) > 0
    );

    pristine.sort((a, b) => {
      // Sort by physical progress DESC, then lowest riskScore ASC, then id ASC
      const progDiff = (b.physicalProgressPercent || 0) - (a.physicalProgressPercent || 0);
      if (progDiff !== 0) return progDiff;
      const rDiff = (a.riskScore ?? 0) - (b.riskScore ?? 0);
      if (rDiff !== 0) return rDiff;
      return a.id.localeCompare(b.id);
    });

    const limit = options.limit || 5;

    return {
      projects: pristine.slice(0, limit),
      totalHealthy: pristine.length,
      qualificationNote: 'Healthy pace qualification requires: PRISM Risk Tier LOW (0–39), 0 months schedule slippage, 0% cost overrun, and positive verified physical progress.'
    };
  }
}

File: backend/services/project_query_service.py
from typing import Optional, List, Dict, Any
from backend.models.project import Project, PortfolioKPIs
from backend.models.common import RiskTier, PriorityTier
from backend.repositories.paimana_repository import paimana_repository

class ProjectQueryService:
    """
    PRISM Canonical Project Query Service.
    Single Source of Truth for querying, filtering, ranking, and aggregating projects
    across Dashboard, Table, Analytics, and AI Copilot.
    """
    @classmethod
    def query(
        cls,
        search: Optional[str] = None,
        state: Optional[str] = None,
        agency: Optional[str] = None,
        sector: Optional[str] = None,
        risk_tier: Optional[str] = None,
        priority_tier: Optional[str] = None,
        sort_by: Optional[str] = None,
        sort_direction: Optional[str] = None,
        limit: Optional[int] = None,
        offset: Optional[int] = None
    ) -> Dict[str, Any]:
        default_sort_by = sort_by or (
            'riskScore' if risk_tier and risk_tier != 'ALL'
            else ('priorityScore' if priority_tier and priority_tier != 'ALL' else 'id')
        )
        default_sort_dir = sort_direction or (
            'desc' if default_sort_by in ('riskScore', 'priorityScore') else 'asc'
        )

        res = paimana_repository.list_projects(
            search=search,
            state=state,
            agency=agency,
            sector=sector,
            risk_tier=risk_tier,
            priority_tier=priority_tier,
            sort_by=default_sort_by,
            sort_direction=default_sort_dir,
            limit=limit,
            offset=offset
        )

        # Compute KPIs on the complete matching dataset, NEVER on paginated slice
        kpis = paimana_repository.compute_kpis(res['allMatching'])

        return {
            'projects': [p.model_dump() for p in res['projects']],
            'allMatching': res['allMatching'],
            'totalCount': res['totalCount'],
            'kpis': kpis.model_dump()
        }

    @classmethod
    def query_structured(
        cls,
        state: Optional[str] = None,
        state_exclude: Optional[List[str]] = None,
        sector: Optional[str] = None,
        sector_exclude: Optional[List[str]] = None,
        risk_tier: Optional[str] = None,
        risk_tier_exclude: Optional[List[str]] = None,
        priority_tier: Optional[str] = None,
        min_risk: Optional[float] = None,
        max_risk: Optional[float] = None,
        min_cost_cr: Optional[float] = None,
        max_cost_cr: Optional[float] = None,
        min_progress_pct: Optional[float] = None,
        max_progress_pct: Optional[float] = None,
        min_delay_months: Optional[int] = None,
        max_delay_months: Optional[int] = None,
        sort_by: Optional[str] = None,
        sort_direction: Optional[str] = None,
        limit: Optional[int] = 5,
        offset: Optional[int] = 0,
        include_multi_state: bool = False
    ) -> Dict[str, Any]:
        """
        Deterministic, adversarial-hardened structured project retrieval engine.
        Supports multi-criteria combinations, negations, numerical bounds,
        dynamic sorting, and deterministic pagination.
        """
        all_projects = paimana_repository.list_projects()['allMatching']
        canonical_state = None
        multi_state_notice: Optional[str] = None

        # 1. State filtering
        if state:
            s = state.strip().lower()
            if s in ('new delhi', 'delhi ncr', 'nct of delhi'):
                s = 'delhi'
            exact_match = next((p for p in all_projects if p.state.strip().lower() == s), None)
            canonical_state = exact_match.state.strip() if exact_match else state.strip()

            dedicated = [p for p in all_projects if p.state.strip().lower() == s]
            multi_state = [
                p for p in all_projects
                if p.state.strip().lower() != s and s in p.state.lower()
            ]

            pool = (dedicated + multi_state) if include_multi_state else dedicated
        else:
            pool = list(all_projects)

        # 2. State exclusion
        if state_exclude:
            normalized_ex = [
                ('delhi' if ex.strip().lower() in ('new delhi', 'delhi ncr', 'nct of delhi') else ex.strip().lower())
                for ex in state_exclude if ex
            ]
            pool = [p for p in pool if not any(ex in p.state.lower() for ex in normalized_ex)]

        # 3. Sector filtering
        if sector:
            sec = sector.strip().lower()
            pool = [
                p for p in pool
                if sec in p.sector.lower() or (p.derivedSector and sec in p.derivedSector.lower())
            ]

        # 4. Sector exclusion
        if sector_exclude:
            ex_secs = [ex.strip().lower() for ex in sector_exclude if ex]
            pool = [
                p for p in pool
                if not any(
                    ex in p.sector.lower() or (p.derivedSector and ex in p.derivedSector.lower())
                    for ex in ex_secs
                )
            ]

        # 5. Risk Tier filtering
        if risk_tier and risk_tier.upper() != 'ALL':
            rt = risk_tier.strip().upper()
            pool = [p for p in pool if p.riskTier == rt]

        # 6. Risk Tier exclusion
        if risk_tier_exclude:
            ex_tiers = {t.strip().upper() for t in risk_tier_exclude if t}
            pool = [p for p in pool if (p.riskTier or 'UNRATED') not in ex_tiers]

        # 7. Priority Tier filtering
        if priority_tier and priority_tier.upper() != 'ALL':
            pt = priority_tier.strip().upper()
            pool = [p for p in pool if p.priorityTier == pt]

        # 8. Numerical Bounds
        if min_risk is not None:
            pool = [p for p in pool if p.riskScore is not None and p.riskScore >= min_risk]

        if max_risk is not None:
            pool = [p for p in pool if p.riskScore is not None and p.riskScore <= max_risk]

        if min_cost_cr is not None:
            pool = [
                p for p in pool
                if (p.revisedCostCr if p.revisedCostCr is not None else (p.originalCostCr or 0.0)) >= min_cost_cr
            ]

        if max_cost_cr is not None:
            pool = [
                p for p in pool
                if (p.revisedCostCr if p.revisedCostCr is not None else (p.originalCostCr or 0.0)) <= max_cost_cr
            ]

        if min_progress_pct is not None:
            pool = [p for p in pool if (p.physicalProgressPercent or 0.0) >= min_progress_pct]

        if max_progress_pct is not None:
            pool = [p for p in pool if (p.physicalProgressPercent or 0.0) <= max_progress_pct]

        if min_delay_months is not None:
            pool = [p for p in pool if (p.timeOverrunMonths or 0) >= min_delay_months]

        if max_delay_months is not None:
            pool = [p for p in pool if (p.timeOverrunMonths or 0) <= max_delay_months]

        # 9. Deterministic Sorting
        sb = (sort_by or '').lower().strip()
        sd = (sort_direction or '').lower().strip()

        # Sensible defaults based on intent
        if not sb:
            if risk_tier == 'LOW' or (max_risk is not None and min_risk is None):
                sb = 'riskscore'
                if not sd:
                    sd = 'asc'
            else:
                sb = 'riskscore'
                if not sd:
                    sd = 'desc'
        elif not sd:
            sd = 'asc' if ('least' in sb or 'low' in sb or 'safest' in sb) else 'desc'

        if sb in ('riskscore', 'risk'):
            if sd == 'asc':
                sort_key = lambda p: ((p.riskScore if p.riskScore is not None else 999), p.id)
            else:
                sort_key = lambda p: (-(p.riskScore if p.riskScore is not None else -1), p.id)
        elif sb in ('cost', 'costcr', 'budget'):
            if sd == 'asc':
                sort_key = lambda p: ((p.revisedCostCr or p.originalCostCr or 0.0), p.id)
            else:
                sort_key = lambda p: (-((p.revisedCostCr or p.originalCostCr or 0.0)), p.id)
        elif sb in ('progress', 'physicalprogress', 'progresspercent'):
            if sd == 'asc':
                sort_key = lambda p: ((p.physicalProgressPercent or 0.0), p.id)
            else:
                sort_key = lambda p: (-(p.physicalProgressPercent or 0.0), p.id)
        elif sb in ('delay', 'timeoverrun', 'overrun'):
            if sd == 'asc':
                sort_key = lambda p: ((p.timeOverrunMonths or 0), p.id)
            else:
                sort_key = lambda p: (-(p.timeOverrunMonths or 0), p.id)
        elif sb in ('priority', 'priorityscore'):
            if sd == 'asc':
                sort_key = lambda p: ((p.priorityScore or 0), p.id)
            else:
                sort_key = lambda p: (-(p.priorityScore or 0), p.id)
        else:
            sort_key = lambda p: (-(p.riskScore if p.riskScore is not None else -1), p.id)

        pool.sort(key=sort_key)

        # Multi-state notice if dedicated state was isolated
        if state and not include_multi_state and canonical_state:
            s_name = canonical_state.strip().lower()
            matching_multi = [
                p for p in all_projects
                if p.state.strip().lower() != s_name and s_name in p.state.lower()
            ]
            if matching_multi:
                corridors = ', '.join(f"{p.name.split('(')[0].strip()} [{p.id}]" for p in matching_multi[:3])
                multi_state_notice = (
                    f"There are also {len(matching_multi)} multi-state project(s) crossing {canonical_state} ({corridors})."
                )

        off = offset or 0
        display_projects = pool[off:off + limit] if limit is not None else pool[off:]

        return {
            'canonicalState': canonical_state,
            'totalCount': len(pool),
            'matchingProjects': pool,
            'displayProjects': display_projects,
            'multiStateNotice': multi_state_notice,
            'appliedFilters': {
                'state': canonical_state,
                'state_exclude': state_exclude,
                'sector': sector,
                'sector_exclude': sector_exclude,
                'risk_tier': risk_tier,
                'risk_tier_exclude': risk_tier_exclude,
                'min_risk': min_risk,
                'max_risk': max_risk,
                'min_cost_cr': min_cost_cr,
                'max_cost_cr': max_cost_cr,
                'min_progress_pct': min_progress_pct,
                'max_progress_pct': max_progress_pct,
                'min_delay_months': min_delay_months,
                'max_delay_months': max_delay_months,
                'sort_by': sb,
                'sort_direction': sd,
                'limit': limit
            }
        }


    @classmethod
    def get_state_partition(cls, state_name: str) -> Optional[Dict[str, Any]]:
        s = state_name.strip().lower()
        all_projects = paimana_repository.list_projects()['allMatching']

        # Find canonical state name
        exact_match = next((p for p in all_projects if p.state.strip().lower() == s), None)
        canonical_state = exact_match.state.strip() if exact_match else state_name.strip()

        dedicated = [p for p in all_projects if p.state.strip().lower() == s]
        multi_state = [
            p for p in all_projects
            if p.state.strip().lower() != s and s in p.state.lower()
        ]

        if not dedicated and not multi_state:
            return None

        all_associated = dedicated + multi_state

        def init_counts() -> Dict[str, int]:
            return {'CRITICAL': 0, 'HIGH': 0, 'MODERATE': 0, 'LOW': 0, 'UNRATED': 0}

        dedicated_by_risk = init_counts()
        for p in dedicated:
            t = p.riskTier or 'UNRATED'
            dedicated_by_risk[t] = dedicated_by_risk.get(t, 0) + 1

        multi_by_risk = init_counts()
        for p in multi_state:
            t = p.riskTier or 'UNRATED'
            multi_by_risk[t] = multi_by_risk.get(t, 0) + 1

        total_by_risk = init_counts()
        for p in all_associated:
            t = p.riskTier or 'UNRATED'
            total_by_risk[t] = total_by_risk.get(t, 0) + 1

        return {
            'canonicalState': canonical_state,
            'dedicatedProjects': dedicated,
            'multiStateProjects': multi_state,
            'allAssociatedProjects': all_associated,
            'dedicatedCount': len(dedicated),
            'multiStateCount': len(multi_state),
            'totalAssociatedCount': len(all_associated),
            'dedicatedKPIs': paimana_repository.compute_kpis(dedicated),
            'allAssociatedKPIs': paimana_repository.compute_kpis(all_associated),
            'dedicatedByRisk': dedicated_by_risk,
            'multiStateByRisk': multi_by_risk,
            'totalByRisk': total_by_risk
        }

    @classmethod
    def get_projects_by_state_and_risk(
        cls,
        state_name: str,
        risk_tier: str,
        limit: int = 5,
        include_multi_state: bool = False,
        sort_ascending: bool = False
    ) -> Dict[str, Any]:
        partition = cls.get_state_partition(state_name)
        target_tier = risk_tier.upper()

        if not partition:
            return {
                'canonicalState': state_name,
                'riskTier': target_tier,
                'dedicatedProjects': [],
                'multiStateProjects': [],
                'totalCount': 0,
                'matchingProjects': [],
                'displayProjects': [],
                'multiStateNotice': None
            }

        # Sort deterministically: ascending for LOW tier (least risky first), descending for others
        if sort_ascending:
            def sort_fn(p: Project):
                return ((p.riskScore if p.riskScore is not None else 999), p.id)
        else:
            def sort_fn(p: Project):
                return (-(p.riskScore if p.riskScore is not None else -1), p.id)

        dedicated_matching = sorted(
            [p for p in partition['dedicatedProjects'] if p.riskTier == target_tier],
            key=sort_fn
        )
        multi_matching = sorted(
            [p for p in partition['multiStateProjects'] if p.riskTier == target_tier],
            key=sort_fn
        )

        canonical_matching = (
            sorted(dedicated_matching + multi_matching, key=sort_fn)
            if include_multi_state else dedicated_matching
        )

        display_projects = canonical_matching[:limit]

        multi_state_notice: Optional[str] = None
        if not include_multi_state and multi_matching:
            corridor_names = ', '.join(f"{p.name.split('(')[0].strip()} [{p.id}]" for p in multi_matching)
            multi_state_notice = (
                f"There are also {len(multi_matching)} multi-state project(s) involving "
                f"{partition['canonicalState']} with {target_tier} risk ({corridor_names})."
            )

        return {
            'canonicalState': partition['canonicalState'],
            'riskTier': target_tier,
            'dedicatedProjects': dedicated_matching,
            'multiStateProjects': multi_matching,
            'totalCount': len(dedicated_matching),
            'matchingProjects': canonical_matching,
            'displayProjects': display_projects,
            'multiStateNotice': multi_state_notice
        }

    @classmethod
    def get_priority_projects(
        cls,
        state: Optional[str] = None,
        sector: Optional[str] = None,
        priority_tier: Optional[str] = None,
        limit: int = 5
    ) -> Dict[str, Any]:
        proj_list = paimana_repository.list_projects()['allMatching']

        if state and state != 'ALL':
            s = state.strip().lower()
            proj_list = [p for p in proj_list if p.state.lower() == s]

        if sector and sector != 'ALL':
            sec = sector.strip().lower()
            proj_list = [
                p for p in proj_list
                if p.sector.lower() == sec or (p.derivedSector or '').lower() == sec
            ]

        if priority_tier and priority_tier != 'ALL':
            pt = priority_tier.strip().upper()
            proj_list = [p for p in proj_list if p.priorityTier == pt]

        # Deterministic sort: priorityScore DESC, riskScore DESC, id ASC
        proj_list.sort(
            key=lambda p: (
                -(p.priorityScore if p.priorityScore is not None else -1),
                -(p.riskScore if p.riskScore is not None else -1),
                p.id
            )
        )

        p1_count = sum(1 for p in proj_list if p.priorityTier == 'P1')
        p2_count = sum(1 for p in proj_list if p.priorityTier == 'P2')
        p3_count = sum(1 for p in proj_list if p.priorityTier == 'P3')

        return {
            'projects': proj_list[:limit],
            'totalMatching': len(proj_list),
            'p1Count': p1_count,
            'p2Count': p2_count,
            'p3Count': p3_count
        }

    @classmethod
    def get_top_risk_projects(
        cls,
        state: Optional[str] = None,
        sector: Optional[str] = None,
        limit: int = 5
    ) -> Dict[str, Any]:
        proj_list = [p for p in paimana_repository.list_projects()['allMatching'] if p.riskScore is not None]

        if state and state != 'ALL':
            s = state.strip().lower()
            proj_list = [p for p in proj_list if p.state.lower() == s]

        if sector and sector != 'ALL':
            sec = sector.strip().lower()
            proj_list = [
                p for p in proj_list
                if p.sector.lower() == sec or (p.derivedSector or '').lower() == sec
            ]

        total_critical = sum(1 for p in proj_list if p.riskTier == 'CRITICAL')
        total_high = sum(1 for p in proj_list if p.riskTier == 'HIGH')

        # Deterministic sort: riskScore DESC, id ASC
        proj_list.sort(key=lambda p: (-(p.riskScore or 0), p.id))

        return {
            'projects': proj_list[:limit],
            'totalCritical': total_critical,
            'totalHigh': total_high
        }

    @classmethod
    def get_stagnant_projects(
        cls,
        state: Optional[str] = None,
        sector: Optional[str] = None,
        limit: int = 5
    ) -> Dict[str, Any]:
        proj_list = paimana_repository.list_projects()['allMatching']

        if state and state != 'ALL':
            s = state.strip().lower()
            proj_list = [p for p in proj_list if p.state.lower() == s]

        if sector and sector != 'ALL':
            sec = sector.strip().lower()
            proj_list = [
                p for p in proj_list
                if p.sector.lower() == sec or (p.derivedSector or '').lower() == sec
            ]

        stagnant = []
        for p in proj_list:
            has_stagnant_driver = (
                (p.primaryRiskDriver and 'stagnat' in p.primaryRiskDriver.lower()) or
                (p.topRiskDrivers and any('stagnat' in d.label.lower() for d in p.topRiskDrivers))
            )
            has_stagnant_alert = (p.alert and p.alert.alertType == 'Progress Stagnation')
            if has_stagnant_driver or has_stagnant_alert:
                stagnant.append(p)

        # Deterministic sort: riskScore DESC, id ASC
        stagnant.sort(key=lambda p: (-(p.riskScore if p.riskScore is not None else -1), p.id))

        return {
            'projects': stagnant[:limit],
            'totalStagnant': len(stagnant)
        }

    @classmethod
    def get_healthy_pace_projects(
        cls,
        state: Optional[str] = None,
        sector: Optional[str] = None,
        limit: int = 5
    ) -> Dict[str, Any]:
        proj_list = paimana_repository.list_projects()['allMatching']

        if state and state != 'ALL':
            s = state.strip().lower()
            proj_list = [p for p in proj_list if p.state.lower() == s]

        if sector and sector != 'ALL':
            sec = sector.strip().lower()
            proj_list = [
                p for p in proj_list
                if p.sector.lower() == sec or (p.derivedSector or '').lower() == sec
            ]

        pristine = [
            p for p in proj_list
            if p.riskTier == 'LOW' and
               (p.timeOverrunMonths or 0) <= 0 and
               (p.costOverrunPercent or 0.0) <= 0.0 and
               (p.physicalProgressPercent or 0.0) > 0.0
        ]

        # Deterministic sort: physicalProgressPercent DESC, lowest riskScore ASC, id ASC
        pristine.sort(
            key=lambda p: (
                -(p.physicalProgressPercent or 0.0),
                (p.riskScore if p.riskScore is not None else 999),
                p.id
            )
        )

        return {
            'projects': pristine[:limit],
            'totalHealthy': len(pristine),
            'qualificationNote': 'Healthy pace qualification requires: PRISM Risk Tier LOW (0–39), 0 months schedule slippage, 0% cost overrun, and positive verified physical progress.'
        }
